File: typescript/src/E4_6_7.ts
// yarn run build && node build/src/E4_6_7.js

type Either<E, T> = Left<E> | Right<T>;

type Left<E> = { tag: 'Left', value: E };
const Left = <E, T>(value: E): Either<E, T> => ({ tag: 'Left', value });
type Right<T> = { tag: 'Right', value: T };
const Right = <E, T>(value: T): Either<E, T> => ({ tag: 'Right', value });

export const map = <E, A, B>(f: (x: A) => B, e: Either<E, A>): Either<E, B> =>
  e.tag === 'Right' ? Right(f(e.value)) : e;

const flatMap = <E, A, B>(f: (x: A) => Either<E, B>, e: Either<E, A>): Either<E, B> =>
  e.tag === 'Right' ? f(e.value) : e;

const orElse = <E, T>(a: Either<E, T>, b: Either<E, T>): Either<E, T> =>
  a.tag === 'Right' ? a : b;

const impossible = () => { throw new Error('impossible') };

const map2 = <E, A, B, C>(f: (x: A, y: B) => C, a: Either<E, A>, b: Either<E, B>): Either<E, C> =>
  a.tag === 'Right' && b.tag === 'Right' ? Right(f(a.value, b.value)) :
  a.tag === 'Left' ? Left(a.value) :
  b.tag === 'Left' ? Left(b.value) :
  impossible();

const traverse = <E, A, B>(f: (x: A) => Either<E, B>, a: A[]): Either<E, B[]> =>
  a.length === 0 ? Right([]) :
    flatMap(hd => flatMap(tl => Right([hd].concat(tl)), traverse(f, a.slice(1))), f(a[0]));

const sequence = <E, T>(e: Either<E, T>[]): Either<E, T[]> =>
  traverse(x => x, e);

File: scala/E4_6_7.scala
// scala E4_6_7.scala

case class Left[+E](value: E) extends Either[E, Nothing]
case class Right[+A](value: A) extends Either[Nothing, A]

sealed trait Either[+E, +A] {

  def map[B](f: A => B): Either[E, B] =
    this match {
      case Left(_) => this.asInstanceOf[Either[E, B]]
      case Right(x) => Right(f(x))
    }

  def flatMap[EE >: E, B](f: A => Either[EE, B]): Either[EE, B] =
    this match {
      case Left(_) => this.asInstanceOf[Either[EE, B]]
      case Right(x) => f(x)
    }

  def orElse[EE >: E,B >: A](b: => Either[EE, B]): Either[EE, B] =
    this match {
      case Left(_) => b
      case Right(_) => this.asInstanceOf[Either[EE, B]]
    }

  def map2[EE >: E, B, C](b: Either[EE, B])(f: (A, B) => C): Either[EE, C] =
    (this, b) match {
      case (Right(x), Right(y)) => Right(f(x, y))
      case (Left(e), _) => this.asInstanceOf[Either[EE, C]]
      case (_, Left(e)) => b.asInstanceOf[Either[EE, C]]
    }

}

object Either {

  def traverse[E, A, B](as: List[A])(f: A => Either[E, B]): Either[E, List[B]] =
    as match {
      case Nil => Right(Nil)
      case x :: xs =>
        for (
          hd <- f(x);
          tl <- traverse(xs)(f)
        ) yield hd :: tl
    }

  def sequence[E, A](es: List[Either[E, A]]): Either[E, List[A]] =
    traverse(es)(x => x)

}
